### PR TITLE
Allow extensions to register git providers

### PR DESCRIPTION
**Pull request by rebornix**
Re https://github.com/Microsoft/vscode-pull-request-github/issues/1030

Since we expose API for other extensions to register their own git provider, we don't have to embed LiveShare logic inside GH PR extension. For example, we can comment out from `extension.ts`.

```
registerLiveShareGitProvider()
```

and then install another extension https://github.com/rebornix/vscode-gh-liveshare , which glues together GH PR and LiveShare. The glue extension declares GH PR and LiveShare as dependencies.

### CODE REVIEW

**Comment by RMacfarlane** [comment]
Hmm, I just tried to submit a review with a comment and got some kind of GraphQL error. I accidentally closed the notification when trying to expand it, so I'm not sure what it was.

Anyway, my summary comment was something like:
The code looks good overall, I'm a little worried about the discoverability about the "glue" extensions. As a user I wouldn't think to search for such a thing, it seems a bit odd.

There is this vscode issue about optional dependencies: https://github.com/Microsoft/vscode/issues/6384

If the GHPR extension was instead able to list LiveShare as an "optional dependency" so that it would get activated later if LiveShare was installed, then it would be possible to just check if LiveShare was enabled and register a git provider for it

File: src/common/utils.ts [hunk 1]
```diff
@@ -196,3 +196,321 @@ export function dateFromNow(date: Date | string): string {
 export interface Predicate<T> {
 	(input: T): boolean;
 }
+
+export class PathIterator implements IKeyIterator {
```
[the remark this answers is not in the document]
Comment (rebornix) [reply src/common/utils.ts]: I should better document this, it's from vscode.

**Comment by rebornix** [comment]
@RMacfarlane thanks for the review and feedback. I happen to have the same discussion with IntelliNav folks about the same topic.

IntelliNav now sets GH PR as an extension dependency so it always attempt to activate GH PR when loaded. It then conflicts with our Nightly build. The workaround we have right now is removing the dependency from IntelliNav and it detects if GH PR or GH PR Nightly is installed or activated or not.

`git provider` is the same story. If an extension wishes GH PR to work in their scenario, they can detect if GH PR exists/activates or not, if so then register proper git providers.

> If the GHPR extension was instead able to list LiveShare as an "optional dependency" so that it would get activated later if LiveShare was installed

what we need here is more like an OR operator for `extensionDependencies`. We want GH PR to be activated when there is Git or IntellINav or LiveShare.

File: src/gitProviders/builtinGit.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+	private _gitAPI: GitAPI;
+	private _disposables: vscode.Disposable[];
+
+	constructor() {
+		const gitExtension = vscode.extensions.getExtension<GitExtension>('vscode.git')!.exports;
+		this._gitAPI = gitExtension.getAPI(1);
+		this._disposables = [];
+		this._disposables.push(this._gitAPI.onDidCloseRepository(e => this._onDidCloseRepository.fire(e as any)));
+		this._disposables.push(this._gitAPI.onDidOpenRepository(e => this._onDidCloseRepository.fire(e as any)));
```
Comment (homitc) [inline src/gitProviders/builtinGit.ts]: Is this meant to be this._onDidOpenRepository?

Comment (RMacfarlane) [reply src/gitProviders/builtinGit.ts]: Yes, definitely. Thanks for catching this!

**Pull request merged**